apps/backend/src/apikeys/apikey.service.ts

```diff
@@ -0,0 +1,87 @@
+import {
+  ForbiddenException,
+  Injectable,
+  NotFoundException
+} from '@nestjs/common';
+import {InjectModel} from '@nestjs/sequelize';
+import {hash} from 'bcryptjs';
+import jwt from 'jsonwebtoken';
+import {CreateApiKeyDto} from '../apikeys/dto/create-apikey.dto';
+import {ConfigService} from '../config/config.service';
+import {User} from '../users/user.model';
+import {UsersService} from '../users/users.service';
+import {ApiKey} from './apikey.model';
+import {APIKeyDto} from './dto/apikey.dto';
+import {UpdateAPIKeyDto} from './dto/update-apikey.dto';
+
+@Injectable()
+export class ApiKeyService {
+  constructor(
+    @InjectModel(ApiKey)
+    private readonly apiKeyModel: typeof ApiKey,
+    private readonly configService: ConfigService,
+    private readonly usersService: UsersService
+  ) {}
+
+  async count(): Promise<number> {
+    return this.apiKeyModel.count();
+  }
+
+  async create(
+    user: User,
+    createApiKeyDto: CreateApiKeyDto
+  ): Promise<{id: string; name: string; apiKey: string}> {
+    const JWTSecret = this.configService.get('JWT_SECRET');
+    const apiKeysAllowed =
+      this.configService.get('API_KEYS_DISABLED')?.toLowerCase() !== 'true';
+    if (apiKeysAllowed && JWTSecret) {
+      const newApiKey = new ApiKey({
+        userId: user.id,
+        name: createApiKeyDto.name
+      });
+      await newApiKey.save();
+      const newJWT = jwt.sign(
+        {keyId: newApiKey.id, createdAt: new Date()},
+        JWTSecret
+      );
+      // Since BCrypt has a 72 byte limit only hash the JWT signature
+      const JWTSignature = newJWT.split('.')[2];
```

Comment:
Can you explain this

Reply:
BCrypt is limited in how many characters it can hash. If the entire JWT token is hashed with BCrypt, only the first 71 characters would be hashed.

The signature is unique to both the payload being signed and the signing key, so hashing the signature should be unique to each JWT token.